src/com/reason/esy/EsyProcess.java

```diff
@@ -154,131 +110,79 @@ public void startNotify() {
 
   private void killIt() {
     if (processHandler == null
-        || processHandler.isProcessTerminating()
-        || processHandler.isProcessTerminated()) {
+            || processHandler.isProcessTerminating()
+            || processHandler.isProcessTerminated()) {
       return;
     }
     processHandler.killProcess();
     processHandler = null;
   }
 
-  private GeneralCommandLine newEsyCommandLine(String command) {
+  private GeneralCommandLine newCommandLine(CliType.Esy cliType) {
     GeneralCommandLine commandLine;
-    if (SystemInfo.isWindows) {
-      commandLine = new GeneralCommandLine("cmd.exe");
-      commandLine.addParameter("/c");
-    } else {
-      commandLine = new GeneralCommandLine("sh");
-      commandLine.addParameter("-c");
-    }
+    commandLine = new GeneralCommandLine(esyExecutable.toString());
     commandLine.setWorkDirectory(workingDir.toFile());
     commandLine.setRedirectErrorStream(redirectErrors);
-    commandLine.addParameter(esyExecutable + " " + command); // 'esy + command' must be a single parameter
+    commandLine.addParameter(getCommand(cliType)); // 'esy + command' must be a single parameter
     return commandLine;
   }
 
-  private Optional<GeneralCommandLine> getDuneCommandLine(CliType.Dune cliType) {
-    Optional<Path> duneExecutable = findDuneExecutableWithinEsy();
-    if (!duneExecutable.isPresent()) {
-      return Optional.empty();
+  private static String getCommand(CliType.Esy cliType) {
+    switch (cliType) {
+      case INSTALL:
+        return Command.INSTALL;
+      case BUILD:
+        return Command.BUILD;
+      case SHELL:
+        return Command.SHELL;
+      default:
+        return Command.ESY;
     }
-
-    String duneCommand = cliType == CliType.Dune.CLEAN ? "clean" : "build"; // @TODO support all actions
-
-    GeneralCommandLine cli = new GeneralCommandLine(duneExecutable.get().toString(), duneCommand);
-    cli.withEnvironment(getEsyEnvironment());
-    cli.setWorkDirectory(workingDir.toString());
-    cli.setRedirectErrorStream(true);
-    return Optional.of(cli);
   }
 
-  private Optional<Path> findDuneExecutableWithinEsy() {
-    Map<String, String> esyEnv = getEsyEnvironment();
-    String paths = esyEnv.get(EnvironmentVariable.PATH);
-    if (paths == null) {
-      return Optional.empty();
-    }
-    return Platform.findExecutableInPath(DUNE_EXECUTABLE_NAME, paths);
+  private static Function<Compiler.ProcessTerminated, ProcessListener> processTerminatedListener =
+          (onProcessTerminated) -> new ProcessAdapter() {
+            @Override
+            public void processTerminated(@NotNull ProcessEvent event) {
+              onProcessTerminated.run();
+            }
+          };
+
+  private static Path findEsyExecutableInPath() {
+    return Platform.findExecutableInPath(ESY_EXECUTABLE_NAME, System.getenv("PATH"))
+            .orElseThrow(() -> {
+              SHOW_ESY_NOT_FOUND_NOTIFICATION.run();
+              return esyNotFoundException();
```

Comment:
is it possible to remove that exception ?

Reply:
changed this up a bit. I got rid of that exception. instead, the process attempts to find the esy executable and working directory each time the process is run. if either are missing, it shows a notification and returns without throwing an exception.